modules/text/src/text_detectorCNN.cpp

```diff
@@ -0,0 +1,458 @@
+        Mat outputMat = Bbox_prob.getMat();
+
+        process_(image.getMat(),outputMat);
+        //copy back to outputArray
+        outputMat.copyTo(Bbox_prob);
+    }
+
+    Size getOutputGeometry()
+    {
+        return this->outputGeometry_;
+    }
+    Size getinputGeometry()
+    {
+        return this->inputGeometry_;
+    }
+
+    int getMinibatchSize()
+    {
+        return this->minibatchSz_;
+    }
+
+    int getBackend()
+    {
+        return OCR_HOLISTIC_BACKEND_DNN;
+    }
+    void setPreprocessor(Ptr<ImagePreprocessor> ptr)
+    {
+        CV_Assert(!ptr.empty());
+        preprocessor_=ptr;
+    }
+
+    Ptr<ImagePreprocessor> getPreprocessor()
+    {
+        return preprocessor_;
+    }
+};
+
+Ptr<DeepCNNTextDetector> DeepCNNTextDetector::create(String archFilename,String weightsFilename,Ptr<ImagePreprocessor> preprocessor,int minibatchSz,int backEnd)
+{
+    if(preprocessor.empty())
+    {
+        // create a custom preprocessor with rawval
+        preprocessor=ImagePreprocessor::createImageCustomPreprocessor(255);
+        // set the mean for the preprocessor
+
+        Mat textbox_mean(1,3,CV_8U);
+        textbox_mean.at<uchar>(0,0)=104;
+        textbox_mean.at<uchar>(0,1)=117;
+        textbox_mean.at<uchar>(0,2)=123;
+        preprocessor->set_mean(textbox_mean);
+    }
+    switch(backEnd){
+    case OCR_HOLISTIC_BACKEND_DEFAULT:
+
+#ifdef HAVE_CAFFE
+        return Ptr<DeepCNNTextDetector>(new DeepCNNTextDetectorCaffeImpl(archFilename, weightsFilename,preprocessor, minibatchSz));
+
+#elif defined(HAVE_DNN)
+        return Ptr<DeepCNNTextDetector>(new DeepCNNTextDetectorDNNImpl(archFilename, weightsFilename,preprocessor, minibatchSz));
+#else
+        CV_Error(Error::StsError,"DeepCNNTextDetector::create backend not implemented");
+        return Ptr<DeepCNNTextDetector>();
+#endif
+    case OCR_HOLISTIC_BACKEND_CAFFE:
+
+        return Ptr<DeepCNNTextDetector>(new DeepCNNTextDetectorCaffeImpl(archFilename, weightsFilename,preprocessor, minibatchSz));
+        break;
+
+    case OCR_HOLISTIC_BACKEND_DNN:
+        return Ptr<DeepCNNTextDetector>(new DeepCNNTextDetectorDNNImpl(archFilename, weightsFilename,preprocessor, minibatchSz));
+        break;
+
+    case OCR_HOLISTIC_BACKEND_NONE:
+    default:
+        CV_Error(Error::StsError,"DeepCNNTextDetector::create backend not implemented");
+        return Ptr<DeepCNNTextDetector>();
+        break;
+    }
+    //return Ptr<DeepCNNTextDetector>();
+
+}
+
+
+Ptr<DeepCNNTextDetector> DeepCNNTextDetector::createTextBoxNet(String archFilename,String weightsFilename,int backEnd)
+{
+
+    // create a custom preprocessor with rawval
+    Ptr<ImagePreprocessor> preprocessor=ImagePreprocessor::createImageCustomPreprocessor(255);
+    // set the mean for the preprocessor
+
+    Mat textbox_mean(1,3,CV_8U);
+    textbox_mean.at<uchar>(0,0)=104;
+    textbox_mean.at<uchar>(0,1)=117;
+    textbox_mean.at<uchar>(0,2)=123;
```

Comment:
where are these magic numbers from? Or how are they calculated? Maybe this should also be part of configuration of this class or read from a prototxt.

Reply:
These numbers are mean values as per the author's original implementation and will be fixed.